hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/util/S3Utils.java

```diff
@@ -131,16 +136,69 @@ public static WebApplicationException wrapOS3Exception(OS3Exception ex) {
             .build());
   }
 
-  public static boolean hasSignedPayloadHeader(HttpHeaders headers) {
-    final String signingAlgorithm = headers.getHeaderString(X_AMZ_CONTENT_SHA256);
-    if (signingAlgorithm == null) {
-      return false;
+  public static boolean hasUnsignedPayload(@Nonnull String amzContentSha256Header) {
+    Objects.requireNonNull(amzContentSha256Header);
+    return amzContentSha256Header.equals(UNSIGNED_PAYLOAD) ||
+        amzContentSha256Header.equals(STREAMING_UNSIGNED_PAYLOAD_TRAILER);
+  }
+
+  public static boolean hasMultiChunksPayload(@Nonnull String amzContentSha256Header) {
+    Objects.requireNonNull(amzContentSha256Header);
+    // Multiple chunk uploads
+    // - https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html
+    // - https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming-trailers.html
+    // Possible values
+    // - STREAMING-UNSIGNED-PAYLOAD-TRAILER
+    // - STREAMING-AWS4-HMAC-SHA256-PAYLOAD
+    // - STREAMING-AWS4-HMAC-SHA256-PAYLOAD-TRAILER
+    // - STREAMING-AWS4-ECDSA-P256-SHA256-PAYLOAD
+    // - STREAMING-AWS4-ECDSA-P256-SHA256-PAYLOAD-TRAILER
+    // Currently since all the multi chunks values have x-amz-content-sha256 header value that starts
+    // with STREAMING, we can use this prefix to differentiates between multi chunks and single chunks upload.
+    // In the future if there are more multi chunks signature algorithms that has the same prefix,
+    // this function will be able to handle detect it.
+    return amzContentSha256Header.startsWith(MULTI_CHUNKS_UPLOAD_PREFIX);
+  }
+
+  public static void validateMultiChunksUpload(HttpHeaders headers, String amzDecodedContentLength,
+                                               String resource) throws OS3Exception {
+    final String contentEncoding = headers.getHeaderString(HttpHeaders.CONTENT_ENCODING);
+    // "Content-Encoding : aws-chunked" seems to only be sent for SDK V2, so ignore if there is no
+    // Content-Encoding header
+    if (contentEncoding == null) {
+      return;
+    }
+    // Amazon S3 supports multiple content encoding values for example "Content-Encoding : aws-chunked,gzip"
+    // We are only interested on "aws-chunked"
+    boolean containsAwsChunked = Arrays.asList(contentEncoding.split(",")).contains(AWS_CHUNKED);
```

Comment:
```suggestion
    boolean containsAwsChunked = Arrays.stream(contentEncoding.split(","))
        .map(String::trim).collect(Collectors.toList()).contains(AWS_CHUNKED);
```

It might be better to add a trim() here. From the examples I saw in the S3 documentation, the Content-Encoding header can have spaces between values, like "Content-Encoding: aws-chunked, gzip".

Reply:
No need to collect results into a list, use `anyMatch`.

```java
Arrays.stream(contentEncoding.split(","))
    .map(String::trim)
    .anyMatch(AWS_CHUNKED::equals);
```

Reply:
Thanks for the suggestion. Updated.